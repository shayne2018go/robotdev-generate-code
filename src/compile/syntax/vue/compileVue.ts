import { ICodeSchema } from '@/types/ICodeSchema';
import compilePages from './compilePages';
import compileRouter, { VueRoute } from './compileRouter';
import compileComponents, { VueComponent } from './compileComponents';
import compileFunctions, { VueFunction } from './compileFunctions';

export interface VueCompileOptions {
  routes: VueRoute[]; // 路由相关
  components: VueComponent[]; // 组件相关
  functions: VueFunction[]; // 函数相关
}

function compileVue(codeSchema: ICodeSchema) {
  const vueCompileOptions = {
    routes: [],
    functions: [],
    components: [],
  } as VueCompileOptions;

  const { tokens: routerTokens, routes } = compileRouter(codeSchema);
  vueCompileOptions['routes'] = routes;

  const { tokens: functionTokens, functions } = compileFunctions(codeSchema);
  vueCompileOptions['functions'] = functions;

  const { tokens: componentTokens, components } = compileComponents(codeSchema, vueCompileOptions);
  vueCompileOptions['components'] = components;

  const { tokens: pageTokens } = compilePages(codeSchema);

  const tokens = routerTokens.concat(componentTokens).concat(functionTokens).concat(pageTokens);

  return { tokens };
}

export default compileVue;
