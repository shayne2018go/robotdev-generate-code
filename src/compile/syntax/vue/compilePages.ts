import createToken from '@/compile/config/createToken';
import { Compile } from '@/types/compile/token';
import { VueGlobalCtx } from './compileVue';
import { BUILT_IN_IMPORTED_PACKAGE, PAGE_DIR } from './const/config';
import { compileScript, compileStyle, compileTemplate } from './sfc';
import { componentEmitsDataStore, componentSlotsDataStore } from './shared/store';
import { nodesDataStore } from './shared/store/nodes';
import { propertiesDataStore } from './shared/store/properties';

export type CompileCurrentCtx = {
  global: VueGlobalCtx;
  scope: {
    current: ParsingCurrentResult;
  };
};

export interface ParsingCurrentResult {
  data: CodeSchema.Component | CodeSchema.Page;
  nodesStore: ReturnType<typeof nodesDataStore>;
  variablesStore: ReturnType<typeof propertiesDataStore>;
  lifeCyclesStore: ReturnType<typeof componentEmitsDataStore>;
  slotsStore: ReturnType<typeof componentSlotsDataStore>;
  emitsStore: ReturnType<typeof componentEmitsDataStore>;
  propsStore: ReturnType<typeof propertiesDataStore>;
  importComponents: GlobalContext.Component[];
  importFunctions: GlobalContext.Function[];
}

// interface ParsingPageResult {
//   page: CodeSchema.Page;
//   nodesStore: ReturnType<typeof nodesDataStore>;
//   variablesStore: ReturnType<typeof propertiesDataStore>;
//   importComponents: GlobalContext.Component[];
//   importFunctions: GlobalContext.Function[];
// }

// var

function compilePages(codeSchema: CodeSchema.Project, vueGlobalCtx: VueGlobalCtx): { tokens: Compile.Token[] } {
  const tokens = [] as Compile.Token[];

  const { directories = [] } = codeSchema;

  codeSchema.pages.forEach((page) => {
    // 1 编译页面代码字符串（token）
    const { token } = compilePage(page, vueGlobalCtx);
    // 2 编译页面地址 ps： 根据directories
    const filePath = getPageFilePath(page, vueGlobalCtx);

    tokens.push(createToken(filePath, token));
  });

  return { tokens };
}

function compilePage(page: CodeSchema.Page, ctx: VueGlobalCtx) {
  const parsingComponentResult = parsingCurrent(page, ctx);

  const currentPageCompileOptions: CompileCurrentCtx = {
    global: ctx,
    scope: {
      current: parsingComponentResult,
    },
  };

  const { token: templateToken } = compileTemplate(page, currentPageCompileOptions);
  const { token: scriptToken } = compileScript(page, currentPageCompileOptions);
  const { token: styleToken } = compileStyle(page, currentPageCompileOptions);

  const token = templateToken + scriptToken + styleToken;

  return { token };
}

export function parsingCurrent<T extends CodeSchema.Component | CodeSchema.Page>(
  data: T,
  ctx: VueGlobalCtx
): ParsingCurrentResult {
  const importComponents: GlobalContext.Component[] = parsingCurrentImportedComponents(data, ctx);

  const buitinImportedFunctions: GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>[] =
    ctx.functionsStore.findAll().filter((func) => {
      return func.source?.packageName && BUILT_IN_IMPORTED_PACKAGE.includes(func.source?.packageName);
    }) || [];

  const importFunctions: GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>[] = [
    {
      id: 'open', // 函数id
      key: 'open', // 函数名称
      source: {
        filePath: 'src/system/action.ts', // 本地路径
        exportName: 'open', // 导出名 (不能为空 默认default )
      },
    } as GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>,
    {
      id: 'queryImage', // 函数id
      key: 'queryImage', // 函数名称
      source: {
        filePath: 'src/system/action.ts', // 本地路径
        exportName: 'queryImage', // 导出名 (不能为空 默认default )
      },
    } as GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>,
    {
      id: 'assignment', // 函数id
      key: 'assignment', // 函数名称
      source: {
        filePath: 'src/system/action.ts', // 本地路径
        exportName: 'assignment', // 导出名 (不能为空 默认default )
      },
    } as GlobalContext.PartialPick<GlobalContext.Function, 'protocol'>,
  ].concat(buitinImportedFunctions);

  const nodesStore = nodesDataStore(data.nodes, ctx);
  const emitsStore = componentEmitsDataStore(isComponent(data) ? data.emits : []);
  const slotsStore = componentSlotsDataStore(isComponent(data) ? data.slots : []);
  const propsStore = propertiesDataStore(isComponent(data) ? data.props : []);
  const variablesStore = propertiesDataStore(data.variables || []);
  const lifeCyclesStore = ctx.eventsStore;

  const result = {
    data,
    nodesStore,
    variablesStore,
    lifeCyclesStore,
    importComponents,
    importFunctions,
    emitsStore,
    slotsStore,
    propsStore,
  } as ParsingCurrentResult;

  return result;
}

function isPage(data: Record<string, any>): data is CodeSchema.Page {
  return data.meta || data.route;
}

function isComponent(data: Record<string, any>): data is CodeSchema.Component {
  return !data.meta && !data.route;
}

function getPageFilePath(page: CodeSchema.Page, vueGlobalCtx: VueGlobalCtx): string {
  // const dirPath: CodeSchema.Directory[] = [];

  // let curId: string | null = page.id;

  // while (curId) {
  //   const current = directories.find((d) => d.id === curId);
  //   if (current) {
  //     dirPath.unshift(current);
  //     curId = current.parentId;
  //   } else {
  //     break;
  //   }
  // }

  // // vue-router 的path
  // const path = `/${dirPath.map((p) => p.key).join('/')}`;

  // // 文件的绝对路径
  // const filePath = `${PAGE_DIR}${path}.vue`;
  const page_protocol = vueGlobalCtx.pagesStore.getPage(page.id);
  if (!page_protocol) {
    throw new Error(`Page ${page.id} is not found`);
  }
  const filePath = page_protocol.source?.filePath || `${PAGE_DIR}/${page_protocol.key}.vue`;

  return filePath;
}

function parsingCurrentImportedComponents<T extends CodeSchema.Component | CodeSchema.Page>(
  data: T,
  ctx: VueGlobalCtx
): GlobalContext.Component[] {
  const importedComponents = [] as GlobalContext.Component[];

  data.nodes.forEach((node) => {
    const isExist = ctx.componentsStore.getCmpt(node.tagId);

    if (isExist && isExist.source?.filePath) {
      importedComponents.push(isExist);
    }
  });
  return importedComponents;
}

export default compilePages;
