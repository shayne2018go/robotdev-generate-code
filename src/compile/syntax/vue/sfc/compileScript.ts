import generate from '@babel/generator';
import * as t from '@babel/types';
import { CompilePageCtx } from '../compilePages';
import { relative } from '@/utils/node';
import { tools } from '@/utils/tools';
import { actionToAst, nodePropValueAst } from '../shared/bind-parse/core';

const lifeCycleMap: { [propname: string]: string } = {
  loading: 'onMounted',
};

function compileScript(page: CodeSchema.Page, ctx: CompilePageCtx): { token: string } {
  return { token: gernateScriptToken(page, ctx) };
}

function gernateScriptToken(page: CodeSchema.Page, ctx: CompilePageCtx): string {
  const { variables, lifeCycle, functions } = page;
  const statements: Array<t.Statement> = [];
  let code = '';
  let tag = getTagStrs();
  code += tag[0];
  statements.push(...getVueImports());
  if (ctx.importComponents.length > 0) {
    statements.push(...getImports(ctx.importComponents));
  }
  if (ctx.importFunctions.length > 0) {
    statements.push(...getImports(ctx.importFunctions));
  }
  if (variables && variables.length > 0) {
    statements.push(getVariables(variables, ctx));
  }
  if (ctx.nodesVarNames) {
    statements.push(getNodesVariables(ctx));
  }
  if (functions && functions.length > 0) {
    statements.push(...getFunctions(functions, ctx));
  }
  if (lifeCycle && lifeCycle.length > 0) {
    statements.push(...getLifeCycles(lifeCycle, ctx));
  }
  const statement = t.program(statements);
  code += generate(statement).code + '\n';
  code += tag[1];
  return code;
}

function getTagStrs(): string[] {
  const statement = t.jsxElement(
    t.jsxOpeningElement(t.jsxIdentifier('script'), [
      t.jsxAttribute(t.jsxIdentifier('setup')),
      t.jsxAttribute(t.jsxIdentifier('lang'), t.stringLiteral('ts')),
    ]),
    t.jsxClosingElement(t.jsxIdentifier('script')),
    [t.jsxText('split')]
  );
  const { code } = generate(statement);
  return code.split('split').map((ele) => ele + '\n');
}

function getVueImports(): t.ImportDeclaration[] {
  return [
    t.importDeclaration(
      [
        t.importSpecifier(t.identifier('onBeforeUnmount'), t.identifier('onBeforeUnmount')),
        t.importSpecifier(t.identifier('onMounted'), t.identifier('onMounted')),
        t.importSpecifier(t.identifier('reactive'), t.identifier('reactive')),
        t.importSpecifier(t.identifier('ref'), t.identifier('ref')),
      ],
      t.stringLiteral('vue')
    ),
    t.importDeclaration(
      [t.importSpecifier(t.identifier('useRouter'), t.identifier('useRouter'))],
      t.stringLiteral('vue-router')
    ),
  ];
}

function getImports(imports: Array<VueTypes.Component> | Array<VueTypes.Function>): t.ImportDeclaration[] {
  const importArray: any[] = [];
  const packageObj: { [propname: string]: number } = {};
  let count = 0;
  imports.forEach((ele) => {
    let { key, source } = ele;
    if (key && source) {
      const { filePath, packageName, exportName, alias } = source;
      const sourceStr = packageName || relative('src/folder', filePath as string);
      if (packageObj[sourceStr] === undefined) {
        let specifier = getImportSpecifier(exportName, key, alias);
        let source = t.stringLiteral(sourceStr);
        importArray.push([[specifier], source]);
      } else {
        importArray[packageObj[sourceStr]][0].push(getImportSpecifier(exportName, key, alias));
      }
      packageObj[sourceStr] = count;
      count++;
    }
  });
  return importArray.map((ele) => {
    return t.importDeclaration(ele[0], ele[1]);
  });
}

function getImportSpecifier(
  exportName: string,
  key: string,
  alias?: string
): t.ImportSpecifier | t.ImportDefaultSpecifier {
  let specifier;
  if (exportName === 'default') {
    specifier = t.importDefaultSpecifier(t.identifier(tools.string.lineToHumpBig(key)));
  } else {
    if (!alias) {
      specifier = t.importSpecifier(t.identifier(exportName), t.identifier(exportName));
    } else {
      specifier = t.importSpecifier(t.identifier(exportName), t.identifier(alias));
    }
  }
  return specifier;
}

function getVariables(variables: Array<CodeSchema.Property_Protocol>, ctx: CompilePageCtx): t.VariableDeclaration {
  return t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(ctx.variablesRootName),
      t.callExpression(t.identifier('reactive'), [
        t.objectExpression(
          variables.map((ele) => {
            return t.objectProperty(t.identifier(ctx.variablesNames[ele.id].varName), t.nullLiteral());
          })
        ),
      ])
    ),
  ]);
}

function getNodesVariables(ctx: CompilePageCtx): t.VariableDeclaration {
  const nodes = ctx.nodesStore.nodes();
  const nodeProps: t.ObjectProperty[] = [];
  nodes.forEach((node) => {
    const propProps: t.ObjectProperty[] = [];
    const eventProps: t.ObjectProperty[] = [];
    if (!node.props?.length && !node.events?.length) {
      return;
    }
    if (node.props?.length) {
      node.props.forEach((prop) => {
        const ast = nodePropValueAst(node.id, prop.propId, ctx);
        if (ast) {
          propProps.push(
            t.objectProperty(t.identifier(ctx.nodesVarNames[node.id].propMembers[prop.propId].varName), ast)
          );
        }
      });
    }
    if (node.events?.length) {
      node.events.forEach((event) => {
        const ast = nodePropValueAst(node.id, event.eventId, ctx);
        if (ast) {
          propProps.push(
            t.objectProperty(t.identifier(ctx.nodesVarNames[node.id].eventMembers[event.eventId].varName), ast)
          );
        }
      });
    }
    nodeProps.push(
      t.objectProperty(
        t.identifier(ctx.nodesVarNames[node.id].varName),
        t.objectExpression([...propProps, ...eventProps])
      )
    );
  });
  return t.variableDeclaration('const', [
    t.variableDeclarator(
      t.identifier(ctx.nodesVarRootName),
      t.callExpression(t.identifier('reactive'), [t.objectExpression(nodeProps)])
    ),
  ]);
}

function getFunctions(functions: Array<CodeSchema.Function_Protocol>, ctx: CompilePageCtx): t.FunctionDeclaration[] {
  return functions.map((func) => {
    return t.functionDeclaration(
      t.identifier(func.key),
      func.parameters.map((param) => {
        return t.identifier(param.key);
      }),
      t.blockStatement([])
    );
  });
}

function getLifeCycles(lifeCycles: Array<CodeSchema.ComponentLifeCycle>, ctx: CompilePageCtx): t.Statement[] {
  return lifeCycles.map((lifeCycle) =>
    t.expressionStatement(
      t.callExpression(t.identifier(lifeCycleMap[ctx.eventsStore.getEvent(lifeCycle.eventId).key]), [
        t.arrowFunctionExpression(
          [],
          t.blockStatement(lifeCycle.actions.map((action) => t.expressionStatement(actionToAst(action, ctx))))
        ),
      ])
    )
  );
}

export default compileScript;
