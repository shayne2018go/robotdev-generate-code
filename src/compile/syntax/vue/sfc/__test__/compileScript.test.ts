import codeSchema from '@/__test__/__fixture__/CodeSchema';
import { describe, expect, it } from 'vitest';
import { CompilePageCtx, parsingCurrent } from '../../compilePages';
import { buildGlobalCtx, parsingVueCompileOptions } from '../../compileVue';
import compileScript from '../compileScript';
import scriptToken from './__fixture__/script-tokens.json';

describe('compileScript', () => {
  // 解析相关依赖协议
  const vueCompileOptions = parsingVueCompileOptions(codeSchema);

  // 构建全局上下文
  const vueGlobalCtx = buildGlobalCtx(vueCompileOptions);

  const page = codeSchema.pages[3];

  const parsingPageResult = parsingCurrent(page, vueGlobalCtx);

  const currentPageCompileOptions: CompilePageCtx = {
    global: vueGlobalCtx,
    scope: {
      current: parsingPageResult,
    },
  };

  it('script 变量 事件', () => {
    const { token } = compileScript(page, currentPageCompileOptions);
    debugger;
    expect(token).toMatchSnapshot();
    expect(token).toEqual(scriptToken);
  });

  it('script 循环 判断', () => {
    const page = codeSchema.pages[0];

    const parsingPageResult = parsingCurrent(page, vueGlobalCtx);

    const currentPageCompileOptions: CompilePageCtx = {
      global: vueGlobalCtx,
      scope: {
        current: parsingPageResult,
      },
    };

    const { token } = compileScript(page, currentPageCompileOptions);
    debugger;

    expect(token).toMatchSnapshot();
  });
});
