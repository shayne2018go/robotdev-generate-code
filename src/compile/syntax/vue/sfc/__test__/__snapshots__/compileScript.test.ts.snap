// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compileScript > script 变量 事件 1`] = `"<script setup lang=\\"ts\\">import{onBeforeUnmount,onMounted,reactive,ref,computed}from\\"vue\\";import{useRouter}from\\"vue-router\\";import{chuangjianyonghu}from\\"../apis/chuangjianyonghu.ts\\";import{chuangjianbanji}from\\"../apis/chuangjianbanji.ts\\";import{luruxueshengbanji}from\\"../apis/luruxueshengbanji.ts\\";import{chaxunxueshengbanji}from\\"../apis/chaxunxueshengbanji.ts\\";import{chaxunsuoyouyonghu}from\\"../apis/chaxunsuoyouyonghu.ts\\";import{chaxunyonghu}from\\"../apis/chaxunyonghu.ts\\";const router=useRouter();const variables=reactive({xueshengmingcheng:null,xueshengnianling:null});const apiState=reactive({chuangjianyonghu:null,chuangjianbanji:null,luruxueshengbanji:null,chaxunxueshengbanji:null,chaxunsuoyouyonghu:null,chaxunyonghu:null});const nodesState=reactive({aCard:{},tpl:{},aForm:{},tpl_2:{},aFormItem:{label:\\"\\\\u5B66\\\\u751F\\\\u540D\\\\u79F0\\"},tpl_3:{},aInput:{value:\\"\\",change:(event_event,event_ctx)=>{variables.xueshengmingcheng={type:\\"data\\",mode:\\"getEventData\\",args:{id:\\"642b8a819a04dc6fabaa5d55\\",path:[\\"642b8a909a04dc6fabaa5d56\\",\\"642be22b451c381a2ec381f7\\"]}}}},aFormItem_2:{label:\\"\\\\u5B66\\\\u751F\\\\u5E74\\\\u9F84\\"},tpl_4:{},aInputNumber:{change:(event_value,event_ctx)=>{variables.xueshengnianling={type:\\"data\\",mode:\\"custom\\",args:{type:\\"number\\",value:121}}}},aFormItem_3:{label:\\"\\\\u5B66\\\\u751F\\\\u5E74\\\\u9F84\\"},tpl_5:{},aButton:{type:\\"primary\\",size:\\"middle\\",shape:\\"round\\"},tpl_6:{},text:{text:\\"\\\\u63D0\\\\u4EA4\\\\u8868\\\\u5355\\"}});</script>"`;

exports[`compileScript > script 循环 判断 1`] = `"<script setup lang=\\"ts\\">import{onBeforeUnmount,onMounted,reactive,ref,computed}from\\"vue\\";import{useRouter}from\\"vue-router\\";import{chuangjianyonghu}from\\"../apis/chuangjianyonghu.ts\\";import{chuangjianbanji}from\\"../apis/chuangjianbanji.ts\\";import{luruxueshengbanji}from\\"../apis/luruxueshengbanji.ts\\";import{chaxunxueshengbanji}from\\"../apis/chaxunxueshengbanji.ts\\";import{chaxunsuoyouyonghu}from\\"../apis/chaxunsuoyouyonghu.ts\\";import{chaxunyonghu}from\\"../apis/chaxunyonghu.ts\\";const router=useRouter();const variables=reactive({xueshengmingcheng:null,xueshengnianling:null,xunhuancols:null});const apiState=reactive({chuangjianyonghu:null,chuangjianbanji:null,luruxueshengbanji:null,chaxunxueshengbanji:null,chaxunsuoyouyonghu:null,chaxunyonghu:null});const nodesState=reactive({each:{data:[\\"a\\",\\"b\\",\\"c\\",\\"d\\"]},tpl:{},each_2:{data:computed(()=>variables.xunhuancols)},tpl_2:{},each_3:{data:computed(()=>apiState.chaxunsuoyouyonghu.data.yonghushuzu)},tpl_3:{},cond:{},tpl_4:{},aCard:{},tpl_5:{},text:{},text_2:{},text_3:{},tpl_6:{},text_4:{text:\\"\\\\u5E74\\\\u9F84\\\\u5C0F\\\\u4E8E10----------\\"}});onMounted(()=>{chaxunsuoyouyonghu(\\"hxy\\").then(res=>{apiState.chaxunsuoyouyonghu=res;open(\\"in\\",\\"blank\\",\\"404\\")}).catch(error=>{})});</script>"`;
